### PR TITLE
Implements all refresh token revocation APIs.

### PR DESCRIPTION
**Pull request by bojeil-google**
This includes:
admin.auth.UserRecord.prototype.tokensValidAfterTime:?string (readonly)
admin.auth.Auth.prototype.verifyIdToken(idToken:string, checkRevoked:boolean=}:!Promise<!admin.auth.DecodedIdToken>
admin.auth.Auth.prototype.revokeRefreshTokens(uid: string): Promise<void>

Unit and integration tests also included.

### CODE REVIEW

File: src/auth/user-record.ts [hunk 1]
```diff
@@ -180,6 +181,9 @@ export class UserRecord {
       // Ignore error.
       utils.addReadonlyGetter(this, 'customClaims', undefined);
     }
+    // Convert validSince first to UTC milliseconds and then to UTC date string.
+    utils.addReadonlyGetter(
```
Comment (hiranya911) [inline src/auth/user-record.ts]: It took a while for me to realize what's going on here. Can we expand this a bit?

```
let validAfterTime: string = null;
if (typeof response.validSince !== 'undefined') {
  validAfterTime = parseDate(...);
}
utils.addReadOnlyGetter(this, 'tokensValidAfterTime', validAfterTime);
```

File: src/auth/auth.ts [hunk 2]
```diff
@@ -124,20 +144,48 @@ export class Auth implements FirebaseServiceInterface {
...
       throw new FirebaseAuthError(
         AuthClientErrorCode.INVALID_CREDENTIAL,
         'Must initialize app with a cert credential or set your Firebase project ID as the ' +
         'GCLOUD_PROJECT environment variable to call auth().verifyIdToken().',
       );
     }
-    return this.tokenGenerator_.verifyIdToken(idToken);
+    return this.tokenGenerator_.verifyIdToken(idToken)
+      .then((decodedIdToken: DecodedIdToken) => {
+        // Whether to check if the token was revoked.
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/auth.ts]: Done

File: src/auth/user-record.ts (hunk 1 above)
Comment (bojeil-google) [reply src/auth/user-record.ts]: Done

File: test/unit/auth/auth-api-request.spec.ts [hunk 3]
```diff
@@ -1313,6 +1329,93 @@ describe('FirebaseAuthRequestHandler', () => {
...
+          expect(stub).to.have.been.calledOnce.and.calledWith(
+              host, port, path, httpMethod, requestData, expectedHeaders, timeout);
+        });
+    });
+
+    it('should be rejected given an invalid uid', () => {
+      const expectedError = new FirebaseAuthError(AuthClientErrorCode.INVALID_UID);
+
+      const requestHandler = new FirebaseAuthRequestHandler(mockApp);
+      return requestHandler.revokeRefreshTokens({'localId': uid} as any)
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/unit/auth/auth-api-request.spec.ts]: Done

File: test/unit/auth/auth-api-request.spec.ts [hunk 4]
```diff
@@ -1313,6 +1329,93 @@ describe('FirebaseAuthRequestHandler', () => {
...
+      // Backend returned error.
+      const expectedError = FirebaseAuthError.fromServerError('USER_NOT_FOUND');
+      const expectedServerError = {
+        error: {
+          message: 'USER_NOT_FOUND',
+        },
+      };
+      const requestData = {
+        localId: uid,
+        validSince: Math.ceil((now.getTime() + 2000) / 1000),
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/unit/auth/auth-api-request.spec.ts]: Done

File: test/unit/auth/auth.spec.ts [hunk 5]
```diff
@@ -97,6 +97,29 @@ function getValidUserRecord(serverResponse: any) {
...
 
 
+/**
+ * Generates a mock decoded ID token with the provided parameters.
+ *
+ * @param {string} uid The uid corresponding to the ID token.
+ * @param {Date} authTime The authentication time of the ID token.
+ * @return {DecodedIdToken} The generated decoded ID token.
+ */
+function getDecodedIdToken(uid: string, authTime: Date) {
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/unit/auth/auth.spec.ts]: Done

File: test/unit/auth/auth.spec.ts [hunk 6]
```diff
@@ -229,7 +268,8 @@ describe('Auth', () => {
     });
 
     it('should forward on the call to the token generator\'s verifyIdToken() method', () => {
-      return auth.verifyIdToken(mockIdToken).then(() => {
+      return auth.verifyIdToken(mockIdToken).then((result) => {
+        expect(result).to.deep.equal(decodedIdToken);
         expect(stub).to.have.been.calledOnce.and.calledWith(mockIdToken);
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/unit/auth/auth.spec.ts]: Good idea. Done.

**Comment by bojeil-google** [comment]
Thanks for the prompt review. I think I addressed all the comments. Let me know if I missed anything.

**Pull request merged**